lib/truffle/java/array_java_proxy.rb

```diff
@@ -6,19 +6,65 @@
 # GNU General Public License version 2
 # GNU Lesser General Public License version 2.1
 
-class ArrayJavaProxy < java.lang.Object
-  include Enumerable
-  
-  def each
-    (0...size).each do |i|
-      yield(self[i])
-    end
-    self
+begin
+  Thread.current[:MAKE_PROXY] = true
+  class ArrayJavaProxy < java.lang.Object
```

Comment:
Well in this particular case we don't want a concrete Java class to be created, so setting the thread local makes all the hooks think we're inside a normal proxy creation case. Since both you and @eregon commented on these I've added a comment explaining these two cases. I tried playing with the load order but ensuring the `inherited` and `included` were added reliably to all necessary classes at the end was much less clear.